A27-xds-global-load-balancing.md

```diff
@@ -0,0 +1,432 @@
+and [priority
+fail-over](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/priority),
+but gRPC assumes that the xDS server will update it to redirect traffic
+when this kind of graceful fail-over is needed.
+
+Unlike the grpclb policy, which relied on server-side load reporting to
+the balancer, the EDS policy will perform client-side load reporting
+using the [LRS
+protocol](https://github.com/envoyproxy/data-plane-api/blob/master/envoy/service/load_stats/v2/lrs.proto).
+Note that the EDS policy will not support per-endpoint stats; it will
+report only per-locality stats.  Currently, the EDS policy will report
+only client-side stats; in the future, we will support also reporting
+backend server stats reported to the client via
+[ORCA](https://docs.google.com/document/d/1NSnK3346BkBo1JUU3I9I5NYYnaJZQPt8_Z_XCBCI3uA/edit).
+
+### How gRPC Will Use the xDS API
+
+gRPC does not have exactly the same set of features as Envoy, so there
+are by necessity some small differences between how the two clients use
+the xDS API.  This section documents exactly what gRPC requires from the
+xDS server and what xDS features it does and does not support.
+
+Because there are many fields in the xDS APIs that gRPC will not use in
+its initial version, the gRPC client must in general ignore any fields
+it does not support.  This does mean that if we add support for one of
+these fields in a future version and users enable use of the feature in
+the management server, older clients will continue to ignore the field,
+so they may not actually get the new behavior.  However, the alternative
+is even worse: if we make the client strict (i.e., it will fail if the
+management server populates fields that the client does not yet support),
+then the same scenario will cause the older clients to fail instead of
+just ignoring the new feature.
+
+Note, however, that there are some unavoidable cases where enabling a
+new feature will cause older clients to break.  These specific cases
+will be called out below.
+
+#### Initial Request on the ADS Stream
+
+The first request on the ADS stream will include a Node message that
+identifies the client.  As mentioned above, most of the fields for the Node
+message will be read from the bootstrap file.
+
+#### LDS
+
+The gRPC client will typically start by sending an LDS request for a
+`Listener` resource whose name matches the server name from the target
+URI used to create the gRPC channel (including port suffix if present).
+Note that the server name will also be used in the HTTP/2 ":authority"
+field, so it needs to be in host or host:port syntax.  However, the gRPC
+client will not interpret it in any special way.  This means that the
+xDS server effectively defines what the default port means if the URI
+does not specify a port.
+
+Because we are requesting one specific resource by name, the LDS response
+should include exactly one `Listener`.  However, many existing xDS servers
```

Comment:
There is some confusion for legacy management servers: I am assuming "do not support requesting one specific resource" means "do not honor resource names in the request". Currently we require the management server to always send client newly requested resources, this implies that server needs to honor what the client requests for. Therefore, I doubt if this kind of legacy management servers would still work for us.

Reply:
> I am assuming "do not support requesting one specific resource" means "do not honor resource names in the request".

Correct.

> Currently we require the management server to always send client newly requested resources, this implies that server needs to honor what the client requests for. Therefore, I doubt if this kind of legacy management servers would still work for us.

For LDS and CDS, legacy management servers will always send all resources in every response.  Therefore, even if the server ignores the `resource_names` field, sending a new request will cause the server to send a new response containing all resources, which must by definition include the new one being requested.

We should not have a problem supporting legacy management servers.